core/src/main/java/org/apache/calcite/plan/RelOptUtil.java

```diff
@@ -2199,13 +2199,49 @@ public static boolean equal(
       RelDataType type2,
       Litmus litmus) {
     if (!areRowTypesEqual(type1, type2, false)) {
-      return litmus.fail("Type mismatch:\n{}:\n{}\n{}:\n{}",
+      return litmus.fail("Type mismatch:\n{}:\n{}\n{}:\n{}\ndifference:\n{}",
           desc1, type1.getFullTypeString(),
-          desc2, type2.getFullTypeString());
+          desc2, type2.getFullTypeString(),
+          getFullTypeDifferenceString(type1, type2));
```

Comment:
Is this method used anywhere?

Reply:
Yes. Some classes use the method.

Reply:
It looks like the error message would duplicate full types multiple times: once from `type1.getFullTypeString()` and the second time would be from `getFullTypeDifferenceString`

Reply:
Sometimes users want to know both what the source type or the target type is and their difference.